Drift/Private/DriftEventManager.cpp

```diff
@@ -57,27 +68,99 @@ TStatId FDriftEventManager::GetStatId() const
 
 void FDriftEventManager::FlushEvents()
 {
+    SCOPE_CYCLE_COUNTER(STAT_FlushDriftEvents);
+
     if (eventsUrl.IsEmpty())
     {
         return;
     }
 
     if (pendingEvents.Num() != 0)
     {
-        auto rm = requestManager.Pin();
-        if (rm.IsValid())
+        if (!requestManager.IsValid())
         {
-            UE_LOG(LogDriftEvent, Verbose, TEXT("[%s] Drift flushing %i events..."), *FDateTime::UtcNow().ToString(), pendingEvents.Num());
-
-            FString payload;
-            JsonArchive::SaveObject(pendingEvents, payload);
-
-            pendingEvents.Empty();
-            auto request = rm->Post(eventsUrl, payload);
-            request->Dispatch();
-            
+            UE_LOG(LogDriftEvent, Error, TEXT("Failed to flush events. Request manager is invalid."));
+            return;
         }
+        
+        UE_LOG(LogDriftEvent, Verbose, TEXT("[%s] Drift flushing %i events..."), *FDateTime::UtcNow().ToString(), pendingEvents.Num());
+
+        TWeakPtr<FDriftEventManager> WeakSelf = SharedThis(this);
+        
+        AsyncTask(ENamedThreads::AnyBackgroundThreadNormalTask, [WeakSelf, Events = MoveTemp(pendingEvents)]()
+        {
+            SCOPE_CYCLE_COUNTER(STAT_ProcessDriftEvents);
+
+            UE_LOG(LogDriftEvent, Verbose, TEXT("Switched to background thread for payload JSON to string processing"));
+
+            const auto StartTime = FPlatformTime::Seconds();
+
+            FString Payload;
+            JsonArchive::SaveObject(Events, Payload);
+
+            TArray<uint8> Compressed;
+            bool bUseCompressed = false;
+            const FTCHARToUTF8 Converter(*Payload);
+            const auto UncompressedSize{ Converter.Length() };
+            if (UncompressedSize >= MIN_SIZE_PAYLOAD_TO_COMPRESS)
+            {
+                UE_LOG(LogDriftEvent, Verbose, TEXT("Attempting to compress payload"));
+                
+                Compressed.SetNumUninitialized(UncompressedSize);
+                const auto Uncompressed = reinterpret_cast<const uint8*>(Converter.Get());
+                int32 CompressedSize;
+                const auto CompressionResult = FCompression::CompressMemory(NAME_Gzip, Compressed.GetData(), CompressedSize, Uncompressed, UncompressedSize);
+                if (CompressionResult && CompressedSize < UncompressedSize)
+                {
+                    UE_LOG(LogDriftEvent, Verbose, TEXT("Payload compression size is smaller than uncompressed size. Using compressed payload."));
+                    
+                    Compressed.SetNum(CompressedSize);
+                    bUseCompressed = true;
+                }
+            }
+
+            const auto EndTime = FPlatformTime::Seconds();
+
+            UE_LOG(LogDriftEvent, Verbose, TEXT("Processed '%d' events in '%.3f' seconds"), Events.Num(), EndTime - StartTime);
+
+            AsyncTask(ENamedThreads::GameThread, [WeakSelf, Payload, Compressed, bUseCompressed]()
```

Comment:
Why copy both Payload and Compressed? Can't you do `auto Payload = bUseCompressed ? Payload : Compressed`?

Reply:
Next step could be to move the compression into the request manager instead, so it becomes useful everywhere